### PR TITLE
feat: Implement GetL2BlockNumberAtSpecificL1Block

### LINKED ISSUE #8: Fork a Network Based on L1 Chain
---

**Title:** Fork a Network Based on L1 Chain Block to Ensure L1-L2 Synchronization

- [x] Implement functionality to fork a network using Multiple anvils
- [x] From the L1 chain, derive the appropriate L2 block to fork.

## Description

We need to ensure that the L1 and L2 (potentially more than 1) chains are in sync when we initiate a fork. The most straightforward way to achieve this synchronization is by basing our fork on an L1 block. From this L1 block, we will derive the appropriate L2 block to fork. This approach guarantees that both chains are consistent and reflective of one another in the forked environment.

## Reasoning

L1 and L2 chains are interdependent, with events and data on the L1 chain influencing the state and behavior of the L2 chain. For accurate testing, debugging, and other operational capabilities, it's crucial that a forked environment maintains this synchronization. For this reason, we base our forks on the L1 chain and derive the corresponding L2 block from there.

## Acceptance Criteria

- Anvil can initiate a network fork based on a specified L1 block.
- The derived L2 block post-fork is consistent and synchronized with its L1 counterpart.

### PR DESCRIPTION
**Pull request by roninjin10**
fixes https://github.com/ethereum-optimism/mocktimism/issues/8

Use the L2OutputOracle to get the l2 fork block number from the l1 fork block number. This will be used to derive the l2 fork block numbers from l1

### CODE REVIEW

**Comment by roninjin10** [comment]
Current dependencies on/for this PR:
* main
  * **PR #69** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/mocktimism/69" target="_blank">Graphite</a> 
    * **PR #70** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/mocktimism/70" target="_blank">Graphite</a> 
      * **PR #71** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/mocktimism/71" target="_blank">Graphite</a> 
        * **PR #74** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/mocktimism/74" target="_blank">Graphite</a> 
          * **PR #75** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/mocktimism/75" target="_blank">Graphite</a> 
            * **PR #76** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/mocktimism/76" target="_blank">Graphite</a> 
              * **PR #78** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/mocktimism/78" target="_blank">Graphite</a>  👈
                * **PR #80** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/mocktimism/80" target="_blank">Graphite</a> 
                  * **PR #81** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/mocktimism/81" target="_blank">Graphite</a> 
                    * **PR #82** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/mocktimism/82" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/mocktimism/78?utm_source=stack-comment).

File: services/anvil/fork_test.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	backend.Commit()
+
+	l1BlockNumber := big.NewInt(1) // As an example
+	l2BlockNumber, err := GetL2BlockNumberAtSpecificL1Block(l1BlockNumber, backend, l2OutputOracleAddress)
+	require.NoError(t, err)
+
+	// TODO add proposals
+
+	expectedL2BlockNumber := big.NewInt(42)
+	require.Equal(t, expectedL2BlockNumber, l2BlockNumber)
```
Comment (roninjin10) [inline services/anvil/fork_test.go]:
```suggestion
	// expectedL2BlockNumber := big.NewInt(42)
	// require.Equal(t, expectedL2BlockNumber, l2BlockNumber)
```
Will finish this test after the proposer is added to mocktimism

File: services/anvil/l2_block_number.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	"context"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-bindings/bindings"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// TODO connect this to superchain repo https://github.com/ethereum-optimism/mocktimism/issues/79
+func (a *AnvilService) L2BlockNumber(l1Client bind.ContractBackend, l2OutputOracleAddr common.Address, l1BlockNumber *big.Int) (*big.Int, error) {
```
Comment (hamdiallam) [inline services/anvil/l2_block_number.go:13]: should the l2 output oracle addr be encapsulated in the service as that's non-changing?

**Review by hamdiallam — APPROVED** [review]

File: services/anvil/l2_block_number.go (hunk 2 above)
Comment (roninjin10) [reply services/anvil/l2_block_number.go:13]: l2OutputOracle is on l1 it has a different addy on every l2

**Pull request merged**